### DISCUSSION TITLE
Error in reference

### DISCUSSION THREAD

**Discussion by ltarmcool** [RudolphRiedel/FT800-FT813#46] [Q&A] [answered]
Hi,
I was able to download the example codes for arduino and compile it.
However, when I tried to write a new display sequence in **src.ino**, I got this reference issued. Can you please point me to the right direction? Thank you

Errors: 
Arduino: 1.8.19 (Windows 10), Board: "Arduino Nano, ATmega328P"

C:\Users\lndinh\AppData\Local\Temp\ccihk18t.ltrans0.ltrans.o: In function `setup':

C:\Users\lndinh\Documents\Arduino\src/src.ino:49: undefined reference to `EVE_cmd_dl(unsigned long)'

C:\Users\lndinh\Documents\Arduino\src/src.ino:50: undefined reference to `EVE_cmd_dl(unsigned long)'

C:\Users\lndinh\Documents\Arduino\src/src.ino:51: undefined reference to `EVE_cmd_dl(unsigned long)'

C:\Users\lndinh\Documents\Arduino\src/src.ino:52: undefined reference to `EVE_color_rgb(unsigned long)'

C:\Users\lndinh\Documents\Arduino\src/src.ino:53: undefined reference to `EVE_cmd_text(int, int, int, unsigned int, char const*)'

C:\Users\lndinh\Documents\Arduino\src/src.ino:54: undefined reference to `EVE_cmd_dl(unsigned long)'

C:\Users\lndinh\Documents\Arduino\src/src.ino:55: undefined reference to `EVE_cmd_dl(unsigned long)'

C:\Users\lndinh\Documents\Arduino\src/src.ino:56: undefined reference to `EVE_execute_cmd()'

collect2.exe: error: ld returned 1 exit status

exit status 1

Error compiling for board Arduino Nano.

![new code](https://user-images.githubusercontent.com/62956288/174849793-74f58cb8-c77e-46c7-9ef9-9059cea111b4.PNG)

**Comment by RudolphRiedel** [comment]
Hello,

which one is that exactly?
https://github.com/RudolphRiedel/FT800-FT813/tree/5.x/examples/EVE_Test_Arduino_PlatformIO ?

While this is using the Arduino framework, it is not for the Arduino IDE.
PlatformIO is a build enviroment to build code for a huge variety of targets and different frameworks.
It is typically used with VSCode: https://platformio.org/install/ide?install=vscode

And PlatformIO is downloading the library code from Github when compiling the example.

Ok, lets try something.

I grabbed this folder:
https://github.com/RudolphRiedel/FT800-FT813/tree/5.x/examples/EVE_Test_Arduino_PlatformIO/src

Renamed the folder to EVE_Test_Arduino.
Renamed file EVE_Test.cpp to EVE_Test_Arduino.ino
Copied all the .c, .cpp and .h files from https://github.com/RudolphRiedel/FT800-FT813 to that folder.
Downloaded Arduino IDE 1.8.19 as .zip.
Unpacked it, started arduino.exe - yup, do not like that.
Opened the EVE_Test_Arduino.ino, selected the Nano, compiled.
Ah yes, does not compile since the display has to be selected, no idea how this is done thru the Arduino IDE.
Changed EVE_config.h to have "#define EVE_EVE3_50G" in line 160, now it compiles.

[EVE_Test_Arduino.zip](https://github.com/RudolphRiedel/FT800-FT813/files/8952140/EVE_Test_Arduino.zip)

The pin configuration is in EVE_target.h, line 1608ff in this case with the nano using an AVR:
#define EVE_CS      10
#define EVE_PDN     8

So yes, sorry, while I do support Arduino, I do not support the Arduino-IDE.

**Reply by ltarmcool** [reply]
Hi Rudolph,
Thank you so much for your quick response. I downloaded your EVE_Test_Arduino.zip file and when I tried a new display sequence, I got the same error as before. I think something is not right with the reference for Arduino-IDE. When I start a new display list in EVE_Test_Arduino.ino, I got compiling error i.e it doesn't recognize the function EVE_cmd_dl or EVE_color_rgb, etc....However, the functions all work fine in tft.c. I found a work around by making a function call Hello_World () in tft.h and tft.c, then call that function is EVE_Test_Arduino.ino and that compiles and uploads fine to the Arduino Nano...Not sure why yet, can you please see if you can repeat this error?
I have attached 2 zip files: Working and Not working
[EVE_Test_Arduino_Working.zip](https://github.com/RudolphRiedel/FT800-FT813/files/8952818/EVE_Test_Arduino_Working.zip)
[EVE_Test_Arduino_NOT_Working.zip](https://github.com/RudolphRiedel/FT800-FT813/files/8952833/EVE_Test_Arduino_NOT_Working.zip)

**Comment by ltarmcool** [comment, marked as answer]
Hi Rudolph,
More information: I downloaded PlatformIO and your codes. I create a new project and manually put all of the library files (EVE.h, EVE_commands.h, etc..)  in subFolder Mylib under lib, in src folder I have main.cpp which I copied and pasted your EVE_Test.cpp codes, and tft.c, tft.h, tft_data.c and tft_data.h files. I experienced the same thing, when called function EVE_cmd_dl directly from main.cpp I got that undefined reference error. But I could call that same function in tft.c...Maybe C++ file has difficulty calling C functions?? 
I notice that you have the C++wrapper for Arduino, should I define that some where in your codes? 
Thank you,
Ltarm

**P.S** I used the dev laptop at work and that doesn't have internet, so I didn't use your platformio.ini to pull the source code directly from Github, would this make a different?

**Reply by RudolphRiedel** [reply]
Yes, this is a C++ issue, or rather an issue with mixing C and C++.
I renamed EVE_commands.c to EVE_commands.cpp and tft.c to tft.cpp.
And in tft.h I commented out lines 8 and 16.
Now EVE_Test_Arduino_NOT_Working compiles.

I have no idea why the C++ compiler can not figure this out by itself.

I used to rename the files for Arduino, mostly because it was not possible to call the SPI class from plain C code.
But then I decided to not do this anymore to have only one set of files for both Arduino and non-Arduino targets.
EVE_target.c used to include the code for the Arduino targets and therefore also had to be renamed.
It was ugly.

So, my suggestion, of course, is to follow my example and use a dedicated .c file for all the Display code. :-)
Roll your own, give it a different name, rewrite all the functions, add more functions, whatever you need. :-)

Or put all the code, including TFT.init() in your .ino file and only rename EVE_commands.c.

**Reply by ltarmcool** [reply]
Awesome, I followed your instructions and loaded the codes to Arduino. Compiles and Works!
Thanks so much for your support! I have a better understanding about the codes and compiler now
Cheers
